hazardous/data/_seer.py

```diff
@@ -0,0 +1,213 @@
+    """
+    if rename:
+        df = pd.read_csv(input_path, sep="\t", header=None)
+        df = add_column_names(df)
+    else:
+        df = pd.read_csv(input_path)
+
+    mask = (
+        (df["SEER cause-specific death classification"] != "N/A not seq 0-59")
+        & (
+            df["Reason no cancer-directed surgery"]
+            != "Not performed, patient died prior to recommended surgery"
+        )
+        & (df["Survival months"] != "Unknown")
+    )
+    df = df.loc[mask]
+
+    df_features = preprocess_features(df)
+    df_events = preprocess_events(df)
+
+    df_features = df_features.astype(COLUMN_NAMES_TO_DTYPE)
+
+    return df_features, df_events
+
+
+def add_column_names(X):
+    """Name inplace the columns of the SEER dataset using a column mapping.
+
+    Parameters
+    ----------
+    X : pandas.DataFrame of shape (n_features, n_samples)
+
+    Returns
+    -------
+    X : pandas.DataFrame of shape (n_features, n_samples)
+    """
+    X.rename(INDICES_TO_COLUMN_NAMES, axis=1, inplace=True)
+    return X
+
+
+def preprocess_features(X):
+    """Replace inplace rare categories to reduce the entropy of the input.
+
+    This is achieved by replacing rare categories by either:
+    - The category frequency
+    - Some placeholder
+    - The most frequent category of the longtail, defined by some thresholds.
+
+    Parameters
+    ----------
+    X : pd.DataFrame of shape (n_samples, n_features)
+
+    Returns
+    -------
+    X : pd.DataFrame of shape (n_samples, n_features - 1)
+    """
+    X = X.drop(columns="Patient ID")
+
+    # Equivalent of OrdinalEncoder on frequencies of "Histologic Type ICD-O-3"
+    val_counts = X["Histologic Type ICD-O-3"].value_counts()
+    replace_dict = {}
+    for idx, count in enumerate(val_counts.values):
+        for k in val_counts[val_counts.values == count].index:
+            replace_dict[k] = str(idx)
+
+    X["Histologic Type ICD-O-3"].replace(replace_dict, inplace=True)
+
+    # Replace special missing values
+    X["ER Status Recode Breast Cancer (1990+)"].replace(
+        "Recode not available",
+        "Positive",
+        inplace=True,
+    )
+    X["PR Status Recode Breast Cancer (1990+)"].replace(
+        "Recode not available",
+        "Positive",
+        inplace=True,
+    )
+    X["Summary stage 2000 (1998-2017)"].replace(
+        "Unknown/unstaged",
+        "Localized",
+        inplace=True,
+    )
+    X["Reason no cancer-directed surgery"].replace(
+        "Unknown; death certificate; or autopsy only (2003+)",
+        "Surgery performed",
+        inplace=True,
+    )
+    X["Median household income inflation adj to 2019"].replace(
+        "Unknown/missing/no match/Not 1990-2018",
+        "$75,000+",
+        inplace=True,
+    )
+
+    X.replace("Unknown", None, inplace=True)
+
+    # TODO: specialize this logic between training and inference.
+    # During inference, we should fetch the trained object.
```

Comment:
This kind of logic should definitely not be part of a dataset loader.

We could instead introduce a `SurvtraceSeerTransformer` with a `fit` and a `transform` method if we really wanted to implement this.

Reply:
Whoops, I forgot to remove these. I indeed created SurvTRACE-specific transformers. I'll add these in a subsequent PR.